quill-core/src/main/scala/io/getquill/dsl/DynamicQueryDSL.scala

```diff
@@ -0,0 +1,376 @@
+    def dynamic: DynamicUpdate[T] = DynamicUpdate(q)
+  }
+
+  implicit class ToDynamicActionReturning[T, U](q: Quoted[ActionReturning[T, U]]) {
+    def dynamic: DynamicActionReturning[T, U] = DynamicActionReturning(q)
+  }
+
+  implicit def dynamicUnquote[T](d: DynamicQuery[T]): Query[T] = macro DynamicQueryDslMacro.dynamicUnquote
+
+  implicit def toQuoted[T](q: DynamicQuery[T]): Quoted[Query[T]] = q.q
+  implicit def toQuoted[T](q: DynamicEntityQuery[T]): Quoted[EntityQuery[T]] = q.q
+  implicit def toQuoted[T <: Action[_]](q: DynamicAction[T]): Quoted[T] = q.q
+
+  def dynamicQuery[T](implicit t: ClassTag[T]): DynamicEntityQuery[T] =
+    dynamicQuerySchema(t.runtimeClass.getSimpleName)
+
+  case class DynamicAlias[T](property: Quoted[T] => Quoted[Any], name: String)
+
+  def alias[T](property: Quoted[T] => Quoted[Any], name: String): DynamicAlias[T] = DynamicAlias(property, name)
+
+  sealed trait DynamicSet[T, U]
+
+  case class DynamicSetValue[T, U](property: Quoted[T] => Quoted[U], value: Quoted[U]) extends DynamicSet[T, U]
+  case class DynamicSetEmpty[T, U]() extends DynamicSet[T, U]
+
+  def set[T, U](property: Quoted[T] => Quoted[U], value: Quoted[U]): DynamicSet[T, U] =
+    DynamicSetValue(property, value)
+
+  def setValue[T, U](property: Quoted[T] => Quoted[U], value: U)(implicit enc: Encoder[U]): DynamicSet[T, U] =
+    set[T, U](property, spliceLift(value))
+
+  def setOpt[T, U](property: Quoted[T] => Quoted[U], value: Option[U])(implicit enc: Encoder[U]): DynamicSet[T, U] =
+    value match {
+      case Some(v) => setValue(property, v)
+      case None    => DynamicSetEmpty()
+    }
+
+  def set[T, U](property: String, value: Quoted[U]): DynamicSet[T, U] =
+    set((f: Quoted[T]) => splice(Constant(property)), value)
+
+  def setValue[T, U](property: String, value: U)(implicit enc: Encoder[U]): DynamicSet[T, U] =
+    set(property, spliceLift(value))
+
+  def dynamicQuerySchema[T](entity: String, columns: DynamicAlias[T]*): DynamicEntityQuery[T] = {
+    val aliases =
+      columns.map { alias =>
+
+        @tailrec def path(ast: Ast, acc: List[String] = Nil): List[String] =
+          ast match {
+            case Property(a, name) =>
+              path(a, name :: acc)
+            case _ =>
+              acc
+          }
+
+        PropertyAlias(path(alias.property(splice[T](Ident("v"))).ast), alias.name)
+      }
+    DynamicEntityQuery(splice[EntityQuery[T]](Entity(entity, aliases.toList)))
+  }
+
+  private[this] val nextIdentId = new DynamicVariable(0)
+
+  private[this] def withFreshIdent[R](f: Ident => R): R = {
+    val idx = nextIdentId.value
+    nextIdentId.withValue(idx + 1) {
+      f(Ident(s"v$idx"))
+    }
+  }
+
+  private def dyn[T](ast: Ast): DynamicQuery[T] =
+    DynamicQuery[T](splice[Query[T]](ast))
+
+  private def splice[T](a: Ast) =
+    new Quoted[T] {
+      override def ast = a
+    }
+
+  protected def spliceLift[O](o: O)(implicit enc: Encoder[O]) =
```

Comment:
Are we vunerable to SQL injection attacks with this?

Reply:
no since it'll produce a regular lift/binding var. There are a few method that take string literals that are vulnerable to injection attacks, though. I'll add a note do the readme about it